src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerGenotypingEngine.java

```diff
@@ -256,6 +274,104 @@ public CalledHaplotypes assignGenotypeLikelihoods(final List<Haplotype> haplotyp
+                    ploidy, originalAlleleCount, practicalAlleleCount, trimmedAlleleMapper.keySet()));
+            return trimmedAlleleMapper;
+        }
+    }
+
+    /**
+     * Trim excessive alt alleles when the combination of ploidy and alleles is so large that exhaustively enumerating
+     * all possible genotypes is impractical.
+     * Use the highest haplotype score of each allele to decide which ones to trim away, and in case of tie, use the second best.
+     * @param alleleMapper          original allele to haplotype map to be trimmed
+     * @param desiredNumOfAlleles   desired allele count (ref allele won't be removed)
+     * @return trimmed map of the desired size
+     */
+    @VisibleForTesting
+    static Map<Allele, List<Haplotype>> removeAltAllelesBasedOnHaptypeScores(final Map<Allele, List<Haplotype>> alleleMapper,
+                                                                             final int desiredNumOfAlleles) {
+
+        final PriorityQueue<AlleleScoredByHaplotypeScores> altAlleleMaxPriorityQ = new PriorityQueue<>();
+
+        final Set<Allele> allelesToRetain = new HashSet<>();
+        // populate allelePriorityQ with the relevant information
+        for(final Allele allele : alleleMapper.keySet()){
+
+            if(allele.isReference()){ // collect scores information only on alt alleles; ref allele is never trimmed by this function
+                allelesToRetain.add(allele);
+                continue;
+            }
+
+            final List<Double> hapScores = alleleMapper.get(allele).stream().map(Haplotype::getScore).sorted().collect(Collectors.toList());
+            final Double highestScore = hapScores.get(hapScores.size()-1);
+            final Double secondHighestScore = hapScores.size()>1 ? hapScores.get(hapScores.size()-2) : Double.NEGATIVE_INFINITY;
+
+            altAlleleMaxPriorityQ.add(new AlleleScoredByHaplotypeScores(allele, highestScore, secondHighestScore));
+        }
+
+        while(allelesToRetain.size()<desiredNumOfAlleles){
+            allelesToRetain.add(altAlleleMaxPriorityQ.poll().getAllele());
+        }
+
+        return alleleMapper.entrySet().stream().filter(p -> allelesToRetain.contains(p.getKey()))
+                .collect(Collectors.toMap(p->p.getKey(), p->p.getValue()));
+    }
+
+    /**
+     * A utility class that provides ordering information, given best and second best haplotype scores.
+     * If there's a tie between the two alleles when comparing their best haplotype score, the second best haplotype score
+     * is used for breaking the tie. In the case that one allele doesn't have a second best allele, i.e. it has only one
+     * supportive haplotype, its second best score is set as {@link Double#NEGATIVE_INFINITY}.
+     * TODO: in the extremely unlikely case that two alleles, having the same best haplotype score, neither have a second
```

Comment:
@SHuang-Broad Would it be possible to use the bases in the alleles themselves to sort in the worst case tie?  Two different alleles are going to have different bases by definition, and if they're equally scored it wouldn't matter which is sorted first.